clients/wallet/src/horizon.rs

```diff
@@ -374,79 +458,71 @@ pub(crate) struct HorizonFetcher<C: HorizonClient> {
+	/// This method is LOOKING FORWARD, so the list is in ASCENDING order:
+	/// starting from the oldest ones, or depending on the last cursor
+	async fn fetch_transactions_iter(&self) -> Result<TransactionsResponseIter<C>, Error> {
 		let public_key_encoded = self.vault_account_public_key.to_encoding();
 		let account_id = std::str::from_utf8(&public_key_encoded).map_err(Error::Utf8Error)?;
 
-		if cursor == 0 {
-			// Fetch the first/latest transaction and set it as the new paging token
-			self.client
-				.get_transactions(account_id, self.is_public_network, 0, 1, false)
-				.await
-		} else {
-			// If we have a paging token, fetch the transactions that occurred after our last stored
-			// paging token
-			self.client
-				.get_transactions(
-					account_id,
-					self.is_public_network,
-					cursor,
-					DEFAULT_PAGE_SIZE,
-					true,
-				)
-				.await
-		}
+		let transactions_response = self
+			.client
+			.get_transactions(
+				account_id,
+				self.is_public_network,
+				self.last_cursor,
+				DEFAULT_PAGE_SIZE,
+				true,
+			)
+			.await?;
+
+		let next_page = transactions_response.next_page();
+		let records = transactions_response.records();
+
+		Ok(TransactionsResponseIter { records, next_page, client: self.client.clone() })
 	}
 
 	/// Fetches the transactions from horizon
 	///
 	/// # Arguments
-	/// * `ledger_env_map` -  a list of TransactionEnvelopes and its corresponding ledger it belongs
-	///   to
+	/// * `ledger_env_map` -  list of TransactionEnvelopes and the ledger it belongs to
 	/// * `targets` - helps in filtering out the transactions to save
 	/// * `is_public_network` - the network the transaction belongs to
 	/// * `filter` - logic to save the needed transaction
-	/// * `last_paging_token` - acts as a marker to scroll over sets of transactions
 	pub async fn fetch_horizon_and_process_new_transactions<T: Clone, U: Clone>(
 		&mut self,
 		ledger_env_map: Arc<RwLock<LedgerTxEnvMap>>,
 		issue_map: Arc<RwLock<T>>,
 		memos_to_issue_ids: Arc<RwLock<U>>,
 		filter: impl FilterWith<T, U>,
-		last_paging_token: PagingToken,
-	) -> PagingToken {
-		let res = self.fetch_latest_txs(last_paging_token).await;
-		let transactions = match res {
-			Ok(txs) => txs._embedded.records,
-			Err(e) => {
-				tracing::warn!("Failed to fetch transactions: {:?}", e);
-				Vec::new()
-			},
-		};
-
-		// Define the new latest paging token as the highest paging token of the transactions we
-		// received or the last paging token if we didn't receive any transactions
-		let latest_paging_token =
-			transactions.iter().map(|tx| tx.paging_token).max().unwrap_or(last_paging_token);
+	) -> Result<(), Error> {
+		let mut response_iter = self.fetch_transactions_iter().await?;
 
 		let (issue_map, memos_to_issue_ids) =
 			future::join(issue_map.read(), memos_to_issue_ids.read()).await;
-		for transaction in transactions {
-			let tx = transaction.clone();
 
+		while let Some(tx) = response_iter.next().await {
```

Comment:
Before calling `next()` we also have to handle the 'current' elements that the iterator was initialized with right? If we just handle the items returned by the call to `.next()` we would miss the `records` that you initialized the struct with in line 503.

Reply:
Nevermind I was misunderstanding the logic in the iterator. Looks good then 👍 

But can we maybe rename `response_iter` to `transactions_iter`?